Python/pythonrun.c

```diff
@@ -280,11 +280,45 @@ PyRun_InteractiveOneObjectEx(FILE *fp, PyObject *filename,
     PyObject *main_dict = PyModule_GetDict(main_module);  // borrowed ref
 
     PyObject *res = run_mod(mod, filename, main_dict, main_dict, flags, arena, interactive_src, 1);
+    Py_INCREF(interactive_src);
     _PyArena_Free(arena);
     Py_DECREF(main_module);
     if (res == NULL) {
+        PyThreadState *tstate = _PyThreadState_GET();
+        PyObject *exc = tstate->current_exception;
+        if ((PyObject *)Py_TYPE(exc) == PyExc_SyntaxError) {
+            /* fix "text" attribute */
+            assert(interactive_src != NULL);
+            PyObject *xs = PyUnicode_Splitlines(interactive_src, 1);
+            Py_DECREF(interactive_src);
+            if (xs == NULL) {
+                return -1;
+            }
+            PyObject *ln = PyObject_GetAttr(exc, &_Py_ID(lineno));
+            if (ln == NULL) {
+                Py_DECREF(xs);
+                return -1;
+            }
+            int n = PyLong_AsInt(ln);
+            Py_DECREF(ln);
+            assert(n > 0);
```

Comment:
You still need to check for errors. This is a public API and there are ways to get these exceptions mutated before they are raised so this assert will fail and in release builds it will go through and crash

Reply:
> there are ways to get these exceptions mutated before they are raised

You meant by user code?  If not, those mutations (breaking invariant ``0<lineno<=len(interactive_src.splitlines())``) looks as bugs.

I'll keep this for a while, lets first see what other core devs think about this idea in general.

Reply:
Ah, indeed.  ``raise x`` does the job.

Reply:
This condition might be tested, but I'm not sure if it worth.  So, added a check without tests